lib/generate_letters.ml

```diff
@@ -38,5 +38,104 @@ let rec index_of_letter curr letter = function
+    'Q';
+    'R';
+    'S';
+    'T';
+    'V';
+    'W';
+    'X';
+    'Y';
+    'Z';
+  ]
+
+(* probability weights associated with each number of vowels in a
+   starting hand. For implementer reference.*)
+(**let weights =
+   [ (0, 0.07); (1, 0.25); (2, 0.36); (3, 0.24); (4, 0.07); (5, 0.01) ]*)
+
+(* alias table reference https://www.keithschwarz.com/darts-dice-coins/*)
+(* probability list needed for an alias table*)
+let prob = [ 0.42; 0.92; 0.56; 1.; 0.42; 0.06 ]
+
+(* alias list needed for an alias table*)
+let alias = [ 1; 2; 3; 0; 2; 2 ]
+
+(** [biased prob total] is true for heads and false for tails. This
+    method simulates a biased coin, with a [prob]/[total] chance of
+    being heads. Requires [prob] <= [total].*)
+let biased prob total =
+  let pick = Random.int total in
+  if pick < prob then true else false
+
+(** [index_item ind curr lst] is the item of lst at index ind. The
+    parameter curr is the current index.*)
+let rec index_item curr ind = function
+  | [] -> raise Not_found
+  | h :: t -> if curr = ind then h else index_item (curr + 1) ind t
+
+(** [get_rand lst] is a random element from lst. *)
+let get_rand lst =
+  let ind = Random.int (List.length lst) in
+  index_item 0 ind lst
+
+(** [num_vowels] is the number of vowels a player's starting hand based
+    on a series of probabilities.*)
+let num_vowels () =
+  let side = Random.int 6 in
+  let pick = index_item 0 side prob in
+  let h_or_t = biased (int_of_float pick * 100) 100 in
+  if h_or_t then side else index_item 0 side alias
+
+(** [lst_of size lst repo] is a list of size [size] of elements from
+    [repo]. Requires [lst] to be empty when first applied.*)
+let rec lst_of size lst repo =
+  match lst with
+  | [] -> lst_of size [ get_rand repo ] repo
+  | h :: t ->
+      if List.length t = size - 1 then h :: t
+      else lst_of size (h :: get_rand repo :: t) repo
+
+(** [optimize_start lst] is a letter_deck of length 7 reflecting
+    accurate letter combinations. Probabilities for vowel/consonant
+    frequency determined using http://www.breakingthegame.net/leaves2 *)
+let optimize_start () =
+  let n_vowels = num_vowels () in
+  let n_conson = 7 - n_vowels in
+  lst_of n_vowels [] vowels @ lst_of n_conson [] consonants
+
+(** [realistic_let] is a char, with a 19% chance of being a vowel. *)
+let realistic_let () =
+  if Random.int 26 < 5 then get_rand vowels else get_rand consonants
+
+(** [optimize_let curr goal lst] removes the appropriate letter when
+    players use a letter and refills it with a new letter. The letter
+    has a 19% chance of being a vowel. Requires curr to be the current
+    index as the function progresses (starting at 0) and goal to be the
+    index of the letter to be removed. remove_let : int -> int ->
+    letter_deck -> letter_deck *)
+let rec optimize_let curr goal = function
+  | [] -> start_game []
+  | h :: t ->
+      if curr = goal then realistic_let () :: t
+      else h :: optimize_let (curr + 1) goal t
```

Comment:
instead of operating on an index, i think it would be more convenient for this function to operate on the letter to be removed. and similar to one of the functions above, instead of having a parameter that must always be zero, you can define a recursive inner function that the outer function then calls starting at zero.

finally, i think that a more descriptive name would be something along the lines of `replace_char_biased`

Reply:
That's what I did originally and in an old pr you commented to change it to be removing by index. I think the indexing makes more sense given that a letter can appear multiple times in a letter_deck but lmk if you disagree.

agreed about the inner recursive function, will do